### PR TITLE
Implimented parsley instance selector UI and functionality

### PR DESCRIPTION
**Pull request by MaisimZaman**
Implemented a parsley instance selector functionality with a pop up UI that allows you to select an instance of parsley to connect to.

- - -
This change is [Reviewable](https://reviewable.io/reviews/waterloo-rocketry/omnibus/124)

### CODE REVIEW

**Review by KavinSatheeskumar — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/omnibus/124)* status: 0 of 6 files reviewed, 2 unresolved discussions (waiting on @BluCodeGH, @MaisimZaman, and @Rio6)

___
*[`sinks/dashboard/items/can_sender.py` line 58 at r2](https://reviewable.io/reviews/waterloo-rocketry/omnibus/124#-NeGLx713pLVSi83lEz9:-NeGLx713pLVSi83lEzA:bf71is1) ([raw file](https://github.com/waterloo-rocketry/omnibus/blob/642fbd7d771e56e2f78fcf6646810977f196ff73/sinks/dashboard/items/can_sender.py#L58)):*
> ```Python
> 
>         # sender
>         self.can_sender = Sender()
> ```

self.can_sender if never used, please remove
___
*[`sources/parsley/main.py` line 22 at r2](https://reviewable.io/reviews/waterloo-rocketry/omnibus/124#-NeGM3pl3fa7IS1daZ3r:-NeGM3pl3fa7IS1daZ3s:bm8d81u) ([raw file](https://github.com/waterloo-rocketry/omnibus/blob/642fbd7d771e56e2f78fcf6646810977f196ff73/sources/parsley/main.py#L22)):*
> ```Python
> class SerialCommunicator:
>     def __init__(self, port, baud, timeout):
>         self.port = port
> ```

This is just something we did so that you could test your changes, please undo all of these "pass" changes

File: sources/parsley/main.py [hunk 1]
```diff
@@ -71,7 +75,7 @@ def main():
             last_heartbeat_time = now
             healthy = "Healthy" if time.time() - last_valid_message_time < 1 else "Dead"
             sender.send(HEARTBEAT_CHANNEL, {
-                "id": f"{gethostname()}/{args.format}", "healthy": healthy})
+                "id": f"{gethostname()}/{args.format}/{os.getpid()}", "healthy": healthy})
```
Comment (Rio6) [inline sources/parsley/main.py:78]: I'd wrap this f"{gethostname()}/{args.format}/{os.getpid()}" in a function or variable, and replace all use later on with it.

Also, would port name be better than using pid? I am thinking if for some reason I need to restart parsley, then the dashboard will need to be updated again.

File: sources/parsley/main.py [hunk 2]
```diff
@@ -19,13 +20,16 @@
...
 
     def read(self):
-        return self.serial.read(4096)
+        # return self.serial.read(4096)
+        return b''
 
     def write(self, msg):
-        self.serial.write(msg)
+        # self.serial.write(msg)
+        pass
```
Comment (Rio6) [inline sources/parsley/main.py:32]: are these commented out code intended

File: sources/parsley/main.py [hunk 3]
```diff
@@ -80,18 +84,23 @@ def main():
...
-            print(formatted_msg)  # always print the usb debug style can message
-            # send the can message over the specified port
-            communicator.write(formatted_msg.encode())
-            last_keepalive_time = now
-            time.sleep(0.01)
+            print(msg)
+            # checking parsley instance
+            parsley_instance = msg.payload['parsley']
+            print(parsley_instance)
+            print(f"{gethostname()}/{args.format}/{os.getpid()}")
```
Comment (Rio6) [inline sources/parsley/main.py:91]: don't need to print these here

**Review by Rio6 — CHANGES_REQUESTED** [review]

File: sources/parsley/main.py (hunk 2 above)
Comment (MaisimZaman) [reply sources/parsley/main.py:32]: I believe  so, should I just remove them?

Comment (Rio6) [reply sources/parsley/main.py:32]: how is parsley going to send and receive serial without these?

Comment (MaisimZaman) [reply sources/parsley/main.py:32]: I think we changed the way parsley sends and receives  serial for testing purposes. If you run individual parsley instances on any specific COM. It should work for testing purposes.

Comment (Rio6) [reply sources/parsley/main.py:32]: For testing yeah, but we probably don't want these testing code merged right?

**Review by KavinSatheeskumar — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/omnibus/124)* status: 0 of 6 files reviewed, 6 unresolved discussions (waiting on @BluCodeGH and @MaisimZaman)

___
*[`sinks/dashboard/can_selector.py` line 1 at r2](https://reviewable.io/reviews/waterloo-rocketry/omnibus/124#-NeZrpcj9XiTXeLGHfDg:-NeZrpcj9XiTXeLGHfDh:b-l6zoyd) ([raw file](https://github.com/waterloo-rocketry/omnibus/blob/642fbd7d771e56e2f78fcf6646810977f196ff73/sinks/dashboard/can_selector.py#L1)):*
> ```Python
> 
> ```

Hey, I know this is a pretty annoying request, but me and Jack were thinking about the best thing for the pop up.

In our opinion, it would be best if, once the user clicks CAN in the menu bar, the things that come down is actually just the list of parsley instances to choose from. That way we don't need a separate pop up

File: sources/parsley/main.py (hunk 2 above)
Comment (MaisimZaman) [reply sources/parsley/main.py]: Yeah, I put them back in the recent commit with all the changes

File: sources/parsley/main.py [hunk 4]
```diff
@@ -80,18 +82,22 @@ def main():
...
-                formatted_msg += ',' + ','.join(f"{byte:02X}" for byte in msg_data)
-            formatted_msg += ";" + crc8.crc8(
-                msg_sid.to_bytes(2, byteorder='big') + bytes(msg_data)
-            ).hexdigest().upper()
-            print(formatted_msg)  # always print the usb debug style can message
-            # send the can message over the specified port
-            communicator.write(formatted_msg.encode())
-            last_keepalive_time = now
-            time.sleep(0.01)
+            print(msg)
```
Comment (Rio6) [inline sources/parsley/main.py:85]: does this print need to be here?

**Review by Rio6 — APPROVED** [review]

**Review by BluCodeGH — CHANGES_REQUESTED** [review]
> Reviewed 1 of 6 files at r1, 1 of 5 files at r2, 1 of 3 files at r6, 1 of 1 files at r9, 1 of 1 files at r10, 1 of 1 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/omnibus/124)* status: all files reviewed, 15 unresolved discussions (waiting on @Darth-Raazi, @KavinSatheeskumar, @MaisimZaman, and @Rio6)

___
*[`sinks/dashboard/dashboard.py` line 24 at r9](https://reviewable.io/reviews/waterloo-rocketry/omnibus/124#-NfUG8e79ypF6ehm2xVX:-NfUG8e84O23amH2L-JX:bsrjppo) ([raw file](https://github.com/waterloo-rocketry/omnibus/blob/387d01f8820c43dc6f69b5fd554bd029c1570275/sinks/dashboard/dashboard.py#L24)):*
> ```Python
> from omnibus.util import TickCounter
> from utils import ConfirmDialog, EventTracker
> from PyQt5 import QtWidgets
> ```

Prefer to use `pyqtgraph.Qt.QtWidgets` to be version-agnostic. See just above where we import all the other widgets, lets be consistent and continue that pattern.
___
*[`sinks/dashboard/dashboard.py` line 96 at r11](https://reviewable.io/reviews/waterloo-rocketry/omnibus/124#-NfY2HvR5SB_mZ4HxtbR:-NfY2HvR5SB_mZ4HxtbS:b-dfgs73) ([raw file](https://github.com/waterloo-rocketry/omnibus/blob/32a87dc069584221178a62888ced4c7a316f19fe/sinks/dashboard/dashboard.py#L96)):*
> ```Python
>         self.current_parsley_instances = []
> 
>         self.last_received = time.time()
> ```

Use `time.monotonic()` for something we're computing a difference with
___
*[`sinks/dashboard/dashboard.py` line 251 at r11](https://reviewable.io/reviews/waterloo-rocketry/omnibus/124#-NfY1DgW2KiUQlsIXlpa:-NfY1DgW2KiUQlsIXlpb:bsnjozo) ([raw file](https://github.com/waterloo-rocketry/omnibus/blob/32a87dc069584221178a62888ced4c7a316f19fe/sinks/dashboard/dashboard.py#L251)):*
> ```Python
>             return retval
> 
>         if (time.time() - self.last_received) > 1:
> ```

What is the goal with this if statement? We only set `last_received` when the set of parsley instances changes?
___
*[`sinks/dashboard/dashboard.py` line 256 at r11](https://reviewable.io/reviews/waterloo-rocketry/omnibus/124#-NfY1AE54JRKHksmzwKC:-NfY1AE54JRKHksmzwKD:bc6eb7n) ([raw file](https://github.com/waterloo-rocketry/omnibus/blob/32a87dc069584221178a62888ced4c7a316f19fe/sinks/dashboard/dashboard.py#L256)):*
> ```Python
>             if self.current_parsley_instances != our_lst:
>                 self.can_selector.clear()
>                 self.last_recieved = time.time()
> ```

Typo in variable name.
___
*[`sinks/dashboard/dashboard.py` line 267 at r11](https://reviewable.io/reviews/waterloo-rocketry/omnibus/124#-NfY0wtJ0fbo-Qtfs4sP:-NfY0wtJ0fbo-Qtfs4sQ:b-gudlpr) ([raw file](https://github.com/waterloo-rocketry/omnibus/blob/32a87dc069584221178a62888ced4c7a316f19fe/sinks/dashboard/dashboard.py#L267)):*
> ```Python
>                     if inst == len(our_lst) - 1:
>                         none_action = self.can_selector.addAction("None")
>                         # just using on_can_select as placeholder
> ```

Placeholder?
___
*[`sinks/dashboard/dashboard.py` line 400 at r11](https://reviewable.io/reviews/waterloo-rocketry/omnibus/124#-NfY0z4BDKjmVFoySYz5:-NfY0z4BDKjmVFoySYz6:btxfmt8) ([raw file](https://github.com/waterloo-rocketry/omnibus/blob/32a87dc069584221178a62888ced4c7a316f19fe/sinks/dashboard/dashboard.py#L400)):*
> ```Python
> 
>         MainWindow = QtWidgets.QMainWindow()
>         ui = CanSelectorWindow(lst, update_parsley_instance)
> ```

`CanSelectorWindow` isn't defined? This whole function is dead code right?
___
*[`sinks/dashboard/items/can_sender.py` line 20 at r11](https://reviewable.io/reviews/waterloo-rocketry/omnibus/124#-NfY1t3Q4k7Y5ry7lh5T:-NfY1t3Q4k7Y5ry7lh5U:bdcuptq) ([raw file](https://github.com/waterloo-rocketry/omnibus/blob/32a87dc069584221178a62888ced4c7a316f19fe/sinks/dashboard/items/can_sender.py#L20)):*
> ```Python
> from .dashboard_item import DashboardItem
> from utils import EventTracker
> from omnibus import Sender
> ```

Sender import isn't used?
___
*[`sources/parsley/main.py` line 21 at r11](https://reviewable.io/reviews/waterloo-rocketry/omnibus/124#-NfY1y6O7iNPnTYfN_vX:-NfY1y6O7iNPnTYfN_vY:b-mo6ieh) ([raw file](https://github.com/waterloo-rocketry/omnibus/blob/d48709badbe5b78136fd8c32e7d5d79b97c8c564/sources/parsley/main.py#L21)):*
> ```Python
> class SerialCommunicator:
>     def __init__(self, port, baud, timeout):
>         self.port = port
> ```

This shouldn't have been deleted?

**Comment by MaisimZaman** [comment]
I pushed the requested changes

**Review by BluCodeGH — CHANGES_REQUESTED** [review]
> Reviewed 2 of 3 files at r12, 1 of 1 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/omnibus/124)* status: all files reviewed, 8 unresolved discussions (waiting on @Darth-Raazi, @KavinSatheeskumar, @MaisimZaman, and @Rio6)

___
*[`sinks/dashboard/dashboard.py` line 251 at r11](https://reviewable.io/reviews/waterloo-rocketry/omnibus/124#-NfY1DgW2KiUQlsIXlpa:-NgACTc390xOxdjUZlf6:by9sc4l) ([raw file](https://github.com/waterloo-rocketry/omnibus/blob/32a87dc069584221178a62888ced4c7a316f19fe/sinks/dashboard/dashboard.py#L251)):*
<details><summary><i>Previously, BluCodeGH (BluCode) wrote…</i></summary><blockquote>

What is the goal with this if statement? We only set `last_received` when the set of parsley instances changes?
</blockquote></details>

bump. Please click on the reviewable button to reply to specific comments.

**Comment by MaisimZaman** [comment]
> > Reviewed 2 of 3 files at r12, 1 of 1 files at r13, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/omnibus/124)_ status: all files reviewed, 8 unresolved discussions (waiting on @Darth-Raazi, @KavinSatheeskumar, @MaisimZaman, and @Rio6)
> 
> _[`sinks/dashboard/dashboard.py` line 251 at r11](https://reviewable.io/reviews/waterloo-rocketry/omnibus/124#-NfY1DgW2KiUQlsIXlpa:-NgACTc390xOxdjUZlf6:by9sc4l) ([raw file](https://github.com/waterloo-rocketry/omnibus/blob/32a87dc069584221178a62888ced4c7a316f19fe/sinks/dashboard/dashboard.py#L251)):_
> 
> Previously, BluCodeGH (BluCode) wrote…
> bump. Please click on the reviewable button to reply to specific comments.

I removed self.last_received

**Review by KavinSatheeskumar — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/omnibus/124)* status: 4 of 6 files reviewed, 11 unresolved discussions (waiting on @BluCodeGH, @Darth-Raazi, @MaisimZaman, and @Rio6)

___
*[`sinks/dashboard/dashboard.py` line 38 at r15](https://reviewable.io/reviews/waterloo-rocketry/omnibus/124#-NgWiEbE3upccDBw5jkk:-NgWiEbE3upccDBw5jkl:b88u9my) ([raw file](https://github.com/waterloo-rocketry/omnibus/blob/86cd01f89787247cc5ba799c75b136205abd4029/sinks/dashboard/dashboard.py#L38)):*
> ```Python
> from omnibus import Sender
> 
> import time
> ```

unused import
___
*[`sinks/dashboard/dashboard.py` line 54 at r15](https://reviewable.io/reviews/waterloo-rocketry/omnibus/124#-NgWiKXL43bmXa4nzgo-:-NgWiKXL43bmXa4nzgo0:bkj70nz) ([raw file](https://github.com/waterloo-rocketry/omnibus/blob/86cd01f89787247cc5ba799c75b136205abd4029/sinks/dashboard/dashboard.py#L54)):*
> ```Python
>         self.zoomed = 1.0
>         self.SCROLL_SENSITIVITY = 1/3  # scale down the scrolling sensitivity
>         self.ui_window = None
> ```

pls delete unused variable
___
*[`sinks/dashboard/dashboard.py` line 153 at r15](https://reviewable.io/reviews/waterloo-rocketry/omnibus/124#-NgWij163cJ27DAaVBNU:-NgWij163cJ27DAaVBNV:b9202u0) ([raw file](https://github.com/waterloo-rocketry/omnibus/blob/86cd01f89787247cc5ba799c75b136205abd4029/sinks/dashboard/dashboard.py#L153)):*
> ```Python
>         for i in range(len(registry.get_items())):
>             new_action = add_item_menu.addAction(registry.get_items()[i].get_name())
>             new_action.triggered.connect(create_registry_trigger(i))  # 5
> ```

get rid of useless comment
___
*[`sinks/dashboard/dashboard.py` line 258 at r15](https://reviewable.io/reviews/waterloo-rocketry/omnibus/124#-NgWj8bpC5E0xwvzHgVs:-NgWj8bpC5E0xwvzHgVt:bercjt9) ([raw file](https://github.com/waterloo-rocketry/omnibus/blob/86cd01f89787247cc5ba799c75b136205abd4029/sinks/dashboard/dashboard.py#L258)):*
> ```Python
>             self.can_selector.clear()
>         
> ```

should probs run format.sh

**Review by BluCodeGH — APPROVED** [review]
> Reviewed 1 of 2 files at r15, 1 of 1 files at r16, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/omnibus/124)* status: all files reviewed, 11 unresolved discussions (waiting on @Darth-Raazi, @MaisimZaman, and @Rio6)

___
*[`sinks/dashboard/dashboard.py` line 242 at r16](https://reviewable.io/reviews/waterloo-rocketry/omnibus/124#-NgjJFP62QBwXGQsk9KO:-NgjJFP62QBwXGQsk9KP:be6voc9) ([raw file](https://github.com/waterloo-rocketry/omnibus/blob/dee69e7160725133c48496862ee8a5c4df2b243e/sinks/dashboard/dashboard.py#L242)):*
> ```Python
>         self.parsley_instance = ''
> 
>     def every_second(self, payload, stream):
> ```

`every_second` isn't a good name if its getting called on every single message through the dashboard. I'd also be a bit worried about the performance implications of that but for now seems fine.

**Comment by Darth-Raazi** [comment]
Outdated PR, closing

**Pull request closed without merging**